### DISCUSSION TITLE
Getting data offset

### DISCUSSION THREAD

**Discussion by bradh** [jamesmudd/jhdf#546] [Q&A] [answered]
I'm looking for a way to get the offset (relative to the start of the file) of the data in a dataset, but not seeing this in the API.

The specific case is in an image, where it looks like this (from `h5ls`):

```
OnTask                   Dataset {3024/3024, 4032/4032, 3/3}
    Attribute: CLASS scalar
        Type:      6-byte null-terminated ASCII string
    Attribute: IMAGE_MINMAXRANGE {2}
        Type:      native unsigned char
    Attribute: IMAGE_SUBCLASS scalar
        Type:      16-byte null-terminated ASCII string
    Attribute: IMAGE_VERSION scalar
        Type:      4-byte null-terminated ASCII string
    Attribute: INTERLACE_MODE scalar
        Type:      16-byte null-terminated ASCII string
    Location:  1:800
    Links:     1
    Storage:   36578304 logical bytes, 36578304 allocated bytes, 100.00% utilization
    Type:      native unsigned char
    Address:   1400
```

I'm looking for a way to get what h5ls calls the "Address", which is essentially the first pixel component value.

I do see `Dataset::getAddress()`, but that gives me what h5ls calls the "Location" value (presumably the start of the dataset structure).

The reason why I'm looking for this is to build a HEIF wrapper around an existing HDF5 file containing images. I don't need the data - I'm trying to use it by reference - but I do need to know where to find it.

**Comment by jamesmudd** [comment, marked as answer]
Thanks for the question. I think what you want is `io.jhdf.api.dataset.ContiguousDataset#getDataAddress` you will need to cast the `Dataset` you have to a `ContiguousDataset` as getting the address for the data doesn't make sense for all types of Dataset. You might prefer to call `io.jhdf.api.dataset.ContiguousDataset#getBuffer` to let jHDF give you a `ByteBuffer` I think you want. Let me know if this answers your question.

**Comment by bradh** [comment]
> I think what you want is `io.jhdf.api.dataset.ContiguousDataset#getDataAddress` you will need to cast the `Dataset` you have to a `ContiguousDataset` as getting the address for the data doesn't make sense for all types of Dataset.

Thanks - that worked well. For the `ContiguousDataset` I assume the data is physically present in the file between [getDataAddress(), getDataAddress() + getStorageSize() - 1].

> You might prefer to call `io.jhdf.api.dataset.ContiguousDataset#getBuffer` to let jHDF give you a `ByteBuffer` I think you want. 

I need the address (well, the byte range) to store in another file (like a metadata sidecar file) which the image reader application will use as an index, and treat the HDF5 file as an opaque container of bits.

> Let me know if this answers your question.

It does (unless the assumption above is wrong). 

Thanks for the quick response - much appreciated.
